dex/pair/src/safe_price.rs

```diff
@@ -0,0 +1,239 @@
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+use super::config;
+use crate::{
+    amm,
+    contexts::base::Context,
+    errors::{ERROR_UNKNOWN_TOKEN, ERROR_ZERO_AMOUNT},
+};
+use common_macros::assert;
+
+const MAX_OBSERVATIONS_PER_RECORD: u64 = 10;
+
+type Block = u64;
+
+#[derive(Clone, TopEncode, TopDecode)]
+pub struct CumulativeState<M: ManagedTypeApi> {
+    pub from: Block,
+    pub to: Block,
+    pub num_observations: u64,
+    pub first_token_reserve: BigUint<M>,
+    pub second_token_reserve: BigUint<M>,
+}
+
+impl<M: ManagedTypeApi> Default for CumulativeState<M> {
+    fn default() -> Self {
+        CumulativeState {
+            from: 0,
+            to: 0,
+            num_observations: 0,
+            first_token_reserve: BigUint::zero(),
+            second_token_reserve: BigUint::zero(),
+        }
+    }
+}
+
+impl<M: ManagedTypeApi> CumulativeState<M> {
+    fn new(block: u64, first_reserve: BigUint<M>, second_reserve: BigUint<M>) -> Self {
+        CumulativeState {
+            from: block,
+            to: block,
+            num_observations: 1,
+            first_token_reserve: first_reserve,
+            second_token_reserve: second_reserve,
+        }
+    }
+
+    fn contains_block(&self, block: u64) -> bool {
+        self.from <= block && block <= self.to
+    }
+
+    fn has_observations(&self) -> bool {
+        self.num_observations != 0
+    }
+
+    fn has_max_observations(&self) -> bool {
+        self.num_observations == MAX_OBSERVATIONS_PER_RECORD
+    }
+
+    fn has_half_max_observations(&self) -> bool {
+        self.num_observations == MAX_OBSERVATIONS_PER_RECORD / 2
+    }
+
+    fn update(
+        &mut self,
+        current_block: u64,
+        first_reserve: BigUint<M>,
+        second_reserve: BigUint<M>,
+    ) {
+        if self.has_observations() {
+            let current_weight = self.to - self.from + 1;
+            let new_weight = self.to - current_block;
```

Comment:
is the `to` block inclusive or exclusive?

because if it is inclusive wouldn't it be ` let new_weight = self.to - current_block +1;`

Reply:
Its inclusive. Thats why the new weight doesn't contain it

Reply:
Oh and btw it should be [curent - to) not (to -current]